### PR TITLE
Expose endpoint to dump timelock lock watch internals

### PR DESCRIPTION
**Pull request by ergo14**
## General
**Before this PR**:
There was no way to dump lock watch state in timelock wrappers.

**After this PR**:
TimeLock agent exposes an endpoint that triggers logs of internal lock watch state.

==COMMIT_MSG==
TimeLock agent exposes an endpoint that triggers logs of internal lock watch state
==COMMIT_MSG==

**Priority**:
P1

**Concerns / possible downsides (what feedback would you like?)**:
See PR comments.

**Is documentation needed?**:
No. 

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:
No, just adding a new endpoint.

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:
No.

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:
Yes.

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:
The endpoint is for exceptional conditions / circumstances, it is meant to be called a few times if things go wrong.

**Does this PR need a schema migration?**
No

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
N/A

**What was existing testing like? What have you done to improve it?**:
Logging PR.

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
Changes introduces should only read from data structures and not mutate anything.

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:
N/A

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:
Internal testing.

**Has the safety of all log arguments been decided correctly?**:
Yes. But, please check.

**Will this change significantly affect our spending on metrics or logs?**:
N/A

**How would I tell that this PR does not work in production? (monitors, etc.)**:
Internal testing.

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:
Recall and rollback.

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
No

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
No

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
Easy to rollback.

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java [hunk 1]
```diff
@@ -49,6 +50,12 @@ long lastVersion() {
         return nextSequence - 1;
     }
 
+    // This method is only for one-off diagnostics purposes.
+    LockWatchEvent[] getBufferSnapshot() {
+        // The contents of the buffer are immutable, thus not requiring a deep copy.
+        return Arrays.copyOf(buffer, buffer.length);
```
Comment (ergo14) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java:56]: We are okay getting null values here. Though, maybe we should just convert it to a list of optionals. Hopefully, internal infra does the null check when it tries to render nulls in arrays.

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java [hunk 2]
```diff
@@ -85,6 +97,21 @@ public synchronized void logLockWatchCreated(LockWatches newWatches) {
...
+    @Override
+    public void dumpState() {
+        diagnosticLog.log(logger -> {
+            log.info(
+                    "Dumping lock event log state: {}, {}, {}, {}",
+                    SafeArg.of("logId", logId),
+                    SafeArg.of("lastVersion", slidingWindow.lastVersion()),
+                    UnsafeArg.of("watches", watchesSupplier.get()),
+                    UnsafeArg.of("openLocks", calculateAllOpenLocks()),
+                    UnsafeArg.of("buffer", slidingWindow.getBufferSnapshot()));
```
Comment (ergo14) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:109]: At this layer we cannot tell the difference between do-not-log and unsafe

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java [hunk 3]
```diff
@@ -116,4 +143,16 @@ private Set<LockDescriptor> calculateOpenLocks(RangeSet<LockDescriptor> watchedR
...
                 .collect(Collectors.toSet());
     }
+
+    /**
+     * Returns a set of all currently held locks.
+     * <p>
+     * Note that the set of held locks can be modified during the execution of this method. Therefore, this method is
+     * NOT guaranteed to return a consistent snapshot of the world.
+     */
+    private Set<LockDescriptor> calculateAllOpenLocks() {
```
Comment (ergo14) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:151]: This is likely not all the lock info we need. For that, we already have diagnostic endpoints for the lock service. Upstream services should trigger those in addition to this one!

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java (hunk 2 above)
Comment (ergo14) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:109]: The log line here could get big!

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java [hunk 4]
```diff
@@ -38,6 +43,13 @@
 import java.util.stream.Collectors;
 
 public class LockEventLogImpl implements LockEventLog {
```
Comment (ergo14) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:45]: The intention here is for all things that dumpState does:
- No writes / mutation
- Should not make the concurrent activity slow

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java [hunk 5]
```diff
@@ -38,6 +43,13 @@
...
+    private static final SafeLogger log = SafeLoggerFactory.get(LockEventLogImpl.class);
+
+    // Log at most 1 line every 2 minutes. Diagnostics are expected to be triggered
+    // on exceptional circumstances and a one-off basis. This de-duplicates when we
+    // need diagnostics on large clusters.
+    private static final RateLimitedLogger diagnosticLog = new RateLimitedLogger(log, 1 / 120.0);
+
```
Comment (barisoyoruk) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:52]: Do we specify this at API level?

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java (hunk 1 above)
Comment (barisoyoruk) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java:56]: Should we check this before merging?

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java (hunk 2 above)
Comment (barisoyoruk) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:109]: buffer looks unsafe
openLocks looks unsafe
watches looks unsafe

lastVersion I couldn't find info but looks safe
logId I couldn't find info and I am not sure if it is safe or unsafe

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java [hunk 6]
```diff
@@ -121,4 +143,16 @@ private Set<LockDescriptor> calculateOpenLocks(RangeSet<LockDescriptor> watchedR
                 .filter(watchedRanges::contains)
                 .collect(Collectors.toSet());
     }
+
+    /**
+     * Returns a set of all currently held locks.
+     * <p>
+     * Note that the set of held locks can be modified during the execution of this method. Therefore, this method is
+     * NOT guaranteed to return a consistent snapshot of the world.
```
Comment (jeremyk-91) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:149]: 👍 good!

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java (hunk 1 above)
Comment (ergo14) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java:56]: Yeah, should be part of some RC testing eventually (before adding code that calls this E2E)

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java (hunk 5 above)
Comment (ergo14) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:52]: We could. Though, I would not personally document this further than the PR discussion. This endpoint should be rarely called

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java (hunk 2 above)
Comment (jeremyk-91) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java:109]: - last version is the sequence number of the lock watch event buffer I think, so is definitely safe.
- logId I'm pretty sure comes from the timelock leader UUID, so is safe.

Yeah I think these are right.

**Review by jeremyk-91 — COMMENTED** [review]
👍 checked everything here except the bit you explicitly said we'd do RC testing for, let's do that before we merge

**Pull request merged**